File: tests/helpers_py/policy_bindings.py
# SPDX-License-Identifier: Apache-2.0

"""Agent policy binding helpers for integration tests.

Mirrors tests/helpers/policy-bindings.ts.
"""

from __future__ import annotations

import re
from typing import Any

import httpx

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)


async def _resolve_policy_id(
    management_url: str, headers: dict[str, str], slug_or_id: str
) -> str | None:
    if _UUID_RE.match(slug_or_id):
        return slug_or_id
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.get(
            f"{management_url}/policies/{slug_or_id}", headers=headers
        )
        if not resp.is_success:
            return None
        return resp.json()["id"]


async def list_agent_bindings(
    management_url: str, headers: dict[str, str], agent_id: str
) -> list[dict[str, Any]]:
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.get(
            f"{management_url}/agents/{agent_id}/bindings", headers=headers
        )
        resp.raise_for_status()
        return resp.json().get("items", [])


async def clear_agent_bindings(
    management_url: str, headers: dict[str, str], agent_id: str
) -> None:
    bindings = await list_agent_bindings(management_url, headers, agent_id)
    async with httpx.AsyncClient(timeout=10.0) as client:
        for b in bindings:
            await client.delete(
                f"{management_url}/agents/{agent_id}/bindings/{b['id']}",
                headers=headers,
            )


async def create_agent_binding(
    management_url: str,
    headers: dict[str, str],
    agent_id: str,
    policy_uuid: str,
    direction: str,
    effect: str,
    config: dict[str, Any] | None = None,
    fail_behavior: str | None = None,
) -> None:
    body: dict[str, Any] = {
        "policyId": policy_uuid,
        "direction": direction,
        "effect": effect,
    }
    if config:
        body["config"] = config
    if fail_behavior:
        body["failBehavior"] = fail_behavior
    async with httpx.AsyncClient(timeout=10.0) as client:
        resp = await client.post(
            f"{management_url}/agents/{agent_id}/bindings",
            headers=headers,
            json=body,
        )
        if not resp.is_success:
            raise RuntimeError(
                f"Failed to create binding for {agent_id}: {resp.status_code} {resp.text}"
            )


async def set_agent_policies(
    management_url: str,
    headers: dict[str, str],
    agent_id: str,
    inbound: list[dict[str, Any]],
    outbound: list[dict[str, Any]],
) -> None:
    """Clear existing bindings and recreate from *inbound*/*outbound* arrays."""
    await clear_agent_bindings(management_url, headers, agent_id)
    for b in inbound:
        raw_id = b.get("policyId")
        if not raw_id:
            continue  # skip bindings without a policy ID
        policy_uuid = await _resolve_policy_id(
            management_url, headers, raw_id
        )
        if not policy_uuid:
            raise RuntimeError(f"Policy not found: {raw_id}")
        await create_agent_binding(
            management_url,
            headers,
            agent_id,
            policy_uuid,
            b.get("direction", "inbound"),
            b.get("effect", "block"),
            b.get("config"),
            b.get("failBehavior"),
        )
    for b in outbound:
        raw_id = b.get("policyId")
        if not raw_id:
            continue
        policy_uuid = await _resolve_policy_id(
            management_url, headers, raw_id
        )
        if not policy_uuid:
            raise RuntimeError(f"Policy not found: {raw_id}")
        await create_agent_binding(
            management_url,
            headers,
            agent_id,
            policy_uuid,
            b.get("direction", "outbound"),
            b.get("effect", "block"),
            b.get("config"),
            b.get("failBehavior"),
        )


async def get_agent_policies(
    management_url: str, headers: dict[str, str], agent_id: str
) -> dict[str, list[dict[str, Any]]]:
    """Return bindings grouped into ``{"inbound": [...], "outbound": [...]}``."""
    bindings = await list_agent_bindings(management_url, headers, agent_id)
    inbound: list[dict[str, Any]] = []
    outbound: list[dict[str, Any]] = []
    for b in bindings:
        d = b.get("direction", "")
        if d in ("inbound", "both"):
            inbound.append(b)
        if d in ("outbound", "both"):
            outbound.append(b)
    return {"inbound": inbound, "outbound": outbound}

File: tests/helpers/policy-bindings.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Shared helpers for integration tests that manage agent policy bindings.
 *
 * After the policy hierarchy redesign (v0.17.0), per-agent JSONB policy
 * columns were replaced by a `policy_bindings` table with CRUD endpoints:
 *   GET    /v1/agents/:agentId/bindings
 *   POST   /v1/agents/:agentId/bindings
 *   DELETE  /v1/agents/:agentId/bindings/:bindingId
 *
 * These helpers provide a backward-compatible interface so integration tests
 * can set/get/clear agent bindings without caring about the CRUD details.
 */

interface BindingInput {
  policyId: string;
  level?: string;
  effect?: string;
  direction?: string;
  config?: Record<string, unknown>;
  failBehavior?: string;
}

interface BindingRow {
  id: string;
  policyId: string | null;
  direction: string;
  effect: string;
  config?: Record<string, unknown>;
}

/**
 * Resolve a policy slug to its UUID. Returns null if not found.
 */
async function resolvePolicyId(
  managementUrl: string,
  headers: Record<string, string>,
  slugOrId: string,
): Promise<string | null> {
  // If it looks like a UUID already, return as-is
  if (
    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(
      slugOrId,
    )
  ) {
    return slugOrId;
  }

  // Look up the policy by slug
  const res = await fetch(`${managementUrl}/policies/${slugOrId}`, { headers });
  if (!res.ok) return null;
  const data = (await res.json()) as { id: string };
  return data.id;
}

/**
 * List current agent-level bindings.
 */
async function listAgentBindings(
  managementUrl: string,
  headers: Record<string, string>,
  agentId: string,
): Promise<BindingRow[]> {
  const res = await fetch(`${managementUrl}/agents/${agentId}/bindings`, {
    headers,
  });
  if (!res.ok) {
    throw new Error(`Failed to list bindings for ${agentId}: ${res.status}`);
  }
  const data = (await res.json()) as { items: BindingRow[] };
  return data.items;
}

/**
 * Delete all agent-level bindings for an agent.
 */
async function clearAgentBindings(
  managementUrl: string,
  headers: Record<string, string>,
  agentId: string,
): Promise<void> {
  const bindings = await listAgentBindings(managementUrl, headers, agentId);
  for (const b of bindings) {
    await fetch(`${managementUrl}/agents/${agentId}/bindings/${b.id}`, {
      method: 'DELETE',
      headers,
    });
  }
}

/**
 * Create a single agent-level binding.
 */
async function createAgentBinding(
  managementUrl: string,
  headers: Record<string, string>,
  agentId: string,
  policyUuid: string,
  direction: string,
  effect: string,
  config?: Record<string, unknown>,
  failBehavior?: string,
): Promise<void> {
  const body: Record<string, unknown> = {
    policyId: policyUuid,
    direction,
    effect,
  };
  if (config) body.config = config;
  if (failBehavior) body.failBehavior = failBehavior;

  const res = await fetch(`${managementUrl}/agents/${agentId}/bindings`, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(
      `Failed to create binding for ${agentId}: ${res.status} ${text}`,
    );
  }
}

/**
 * Set agent policies — backward-compatible replacement for the old
 * `PUT /agents/:agentId/policies` endpoint.
 *
 * Clears all existing agent-level bindings, then creates new ones from
 * the provided inbound/outbound arrays.
 */
export async function setAgentPolicies(
  managementUrl: string,
  headers: Record<string, string>,
  agentId: string,
  inbound: BindingInput[],
  outbound: BindingInput[],
): Promise<void> {
  await clearAgentBindings(managementUrl, headers, agentId);

  for (const b of inbound) {
    const policyUuid = await resolvePolicyId(
      managementUrl,
      headers,
      b.policyId,
    );
    if (!policyUuid) {
      throw new Error(`Policy not found: ${b.policyId}`);
    }
    await createAgentBinding(
      managementUrl,
      headers,
      agentId,
      policyUuid,
      b.direction ?? 'inbound',
      b.effect ?? 'block',
      b.config,
      b.failBehavior,
    );
  }

  for (const b of outbound) {
    const policyUuid = await resolvePolicyId(
      managementUrl,
      headers,
      b.policyId,
    );
    if (!policyUuid) {
      throw new Error(`Policy not found: ${b.policyId}`);
    }
    await createAgentBinding(
      managementUrl,
      headers,
      agentId,
      policyUuid,
      b.direction ?? 'outbound',
      b.effect ?? 'block',
      b.config,
      b.failBehavior,
    );
  }
}

/**
 * Get agent policies — backward-compatible replacement for the old
 * `GET /agents/:agentId/policies` endpoint.
 *
 * Returns bindings grouped into inbound/outbound arrays.
 */
export async function getAgentPolicies(
  managementUrl: string,
  headers: Record<string, string>,
  agentId: string,
): Promise<{ inbound: BindingRow[]; outbound: BindingRow[] }> {
  const bindings = await listAgentBindings(managementUrl, headers, agentId);
  const inbound: BindingRow[] = [];
  const outbound: BindingRow[] = [];

  for (const b of bindings) {
    if (b.direction === 'inbound' || b.direction === 'both') {
      inbound.push(b);
    }
    if (b.direction === 'outbound' || b.direction === 'both') {
      outbound.push(b);
    }
  }

  return { inbound, outbound };
}
